File: php/3. Catch Various Book Reference Types.php
<?php 
/*************************************************************************************
 * 
 * 	This script is to demonstrate how to catch a variety of reference types including 
 * 		abbreviations and misspellings.
 *	It will used the function getBookByKeyword which takes what is offered as the 
 *		book, whether it's the full name, an abbreviation or a misspelling and 
 *		return the data contained on the kjv_books database for that book.
 *	We will also use the function putNumRefInArray to take the numeric part of the 
 *		reference and return an array of verses.
 * 
 ***************************************************************************************/

/*******************************
 * 
 * Connect to the database. 
 * For demonstration purposes I actually used db_user and db_password for my local intranet.
 * 	You'll want to use something more secure.
 * The database is named `osbible` (Yours may be different.)
 * 
 *****************************/

  # $_mysql will be our db connection variable. Mysqli requires it for almost everything.
$_mysql = mysqli_connect('localhost','db_user','db_password','osbible');


	# We will start with a few typical passage reference.
	# 	We will put them in anarray because we'll used several of them to demostrate various forms of references.

$References=[];
$Note[1]='A typical abbreviation';
$References[1]='Jn 3:16';
$Note[2]='Several examples of how books with numbers are entered. Just a number';
$References[2]='2 pt 1:1';
$Note[3]='...Number with ordinal';
$References[3]='2nd pt 1:1';
$Note[4]='...Roman numeral';
$References[4]='ii pt 1:1';
$Note[5]='...Ordinal word';
$References[5]='Second Peter 1:1';
$Note[6]='Common mispelling';
$References[6]='gnesis 1:1';

foreach($References as $n=>$reference)
	{
	# We'll reserve the original reference
	$_reference=$reference;
	# We;ll grab the note to display
	$note=$Note[$n];
	# To avoid confusing the filter we'll make all book references lower case
	$reference=strtolower($reference);
	# We'll use the function to split the book part from the nymber part and return it all in an array
	$BookData=getRefByKeyword($reference);
	# We want the book id for the next database query
	$bid=$BookData['id'];
	# Here we'll have the book name
	$book=$BookData['book'];
	# We'll split the chapter from the verse
	list($chapter,$verse)=explode(':',$BookData['num_key']);
	# Now we'll build our query... 
	$querytext=sprintf("SELECT	`text` 		FROM	`kjvs`
						WHERE	`book` 		= 		'%s'
						AND 	`chapter`	=		'%s'
						AND		`verse`		=		'%s'
						LIMIT	1;",
						mysqli_real_escape_string($_mysql, $bid),
						mysqli_real_escape_string($_mysql, $chapter),
						mysqli_real_escape_string($_mysql, $verse));
	# Submit the query
	$query = mysqli_query($_mysql, $querytext);
	# Receive the results
	$Verse = mysqli_fetch_array($query);
	# Assign the data to $text
	$text=$Verse['text'];
	# Filter out the Strong's numbers (we don't need them now)
	$text=preg_replace('/\{(.*?)\}/', '', $text);
	# Echo the original reference with the verse and correct reference
	echo "Original reference: $_reference ($note)<br>Result: &ldquo;$text&rdquo;&mdash;$book $chapter:$verse<hr>";	
	}




function getRefByKeyword($k)
	{
	# Bring our databse connection within the scope of the function
	global $_mysql;
	# Set some variables
	$Return['id']=0;
	$Return['book']='';
	$Return['chapters']=0;
	# Clean any oddities brought over in a url
	$_k=urldecode($k);
	# Remove extra spaces
	$_k=preg_replace("/(\s){2,}/", ' ', $_k); 
	# Remove periods
	$_k=str_replace('.','',$_k);
	# Make ordinals regular numbers (note, here we are expecting a space between the ordinal and the book name)
	$_k=str_replace('1st ', '1 ', $_k);
	$_k=str_replace('2nd ', '2 ', $_k);
	$_k=str_replace('3rd ', '3 ', $_k);
	$_k=str_replace('first ', '1 ', $_k);         
	$_k=str_replace('second ', '2 ', $_k);
	$_k=str_replace('third ', '3 ', $_k);
	$_k=preg_replace('/^i /i', "1 ", $_k); 
	$_k=preg_replace('/^ii /i', "2 ", $_k);
	$_k=preg_replace('/^iii /i', "3 ", $_k);

	# Catch messed up beginning numbers
	# 	This catches initial numbers imediately followed by text, as in 1pe or 2jn
	if(preg_match('/^[1-3][a-zA-Z]/',$_k))
		{
		$_k=preg_replace('/^1/', '1 ',$_k);
		$_k=preg_replace('/^2/', '2 ',$_k);
		$_k=preg_replace('/^3/', '3 ',$_k);
		}
	# Explode by spaces
	$Ref_keys=explode(' ',$_k);
	# If first element is empty, remove (this happens whe the reference begins with a space)
	if(!$Ref_keys[0])
		{
		$toss=array_shift($Ref_keys);
		}
	# If first element is a number, combine with second to make the book key
	if(isset($Ref_keys[1]) and preg_match('/^[1-9]/',$Ref_keys[0]))
		{
		$book_key=$Ref_keys[0].' '.$Ref_keys[1];
		$num_keys=$Ref_keys[2];
		}
	else
		{
		$book_key=$Ref_keys[0];
		$num_keys=$Ref_keys[1];
		}
	# Add the chapter and verse(s) to the output as 'num_key'
	$Return['num_key']=$num_keys;
	# I don't remember why, but I found this necessary
	if(strtolower($book_key)=='jud'){$book_key='Judges';}
	if(strtolower($book_key)=='eph'){$book_key='Ephesians';}
		
	$_c=0;
	# Let's explain the query... 
	# 	First we look in the `abbr` column for matches. Theindividual abbreviations are delimited by | 
	# 	We "score" the macthes so that the one with the best "score" rises to the top and we get it.
    $queryText = sprintf("SELECT * FROM `kjv_books` WHERE `abbr` LIKE '%%|%s|%%' || `book` LIKE '%s' || `kjav_abr` LIKE '%s' || `book` SOUNDS LIKE '%s'
                         ORDER BY
                         case when `abbr` LIKE '%%|%s|%%' then 4 else 0 end
                       + case when `book` LIKE '%s' then 3 else 0 end
                       + case when `kjav_abr` LIKE '%s' then 2 else 0 end
                       + case when `book` SOUNDS LIKE '%s' then 1 else 0 end
                         DESC LIMIT 1",
                 mysqli_real_escape_string($_mysql,$book_key),
                 mysqli_real_escape_string($_mysql,$book_key),
                 mysqli_real_escape_string($_mysql,$book_key),
                 mysqli_real_escape_string($_mysql,$book_key),
                 mysqli_real_escape_string($_mysql,$book_key),
                 mysqli_real_escape_string($_mysql,$book_key),
                 mysqli_real_escape_string($_mysql,$book_key),
                 mysqli_real_escape_string($_mysql,$book_key));
	$query=mysqli_query($_mysql,$queryText);
	if(mysqli_errno($_mysql)){echo ": " . mysqli_error($_mysql) . "<br>$queryText\n<hr>";}
	if(mysqli_num_rows($query))
		{
		$c=0;
		while ($dbRow = mysqli_fetch_assoc($query)) 
			{
			$Return['dbRow']=$dbRow;
			$c++;
			if($dbRow['book']=='Psalms'){$dbRow['book']='Psalm';}
			$Return['id']=$dbRow['id'];
			$Return['book']=$dbRow['book'];
			$Return['chapters']=$dbRow['chapters'];
			}
		}		
	$Return['queryText']=$queryText;
	return $Return;
	}



?>

File: python/3. Catch Various Book Reference Types.py
#!C:/xampp/python/python.exe
"""
This script demonstrates how to catch a variety of reference types including 
abbreviations and misspellings.

It uses the function get_book_by_keyword which takes what is offered as the 
book (whether it's the full name, an abbreviation or a misspelling) and 
returns the data contained in the kjv_books database for that book.

We also use the numeric part of the reference to return verse data.
"""



#import cgi
import pymysql
import re
from urllib.parse import unquote



# Print CGI header
print("Content-Type: text/html\n")


# Connect to the database
try:
    conn = pymysql.connect(
        host='localhost',
        user='db_user',
        password='db_password',
        database='osbible',
        cursorclass=pymysql.cursors.DictCursor
    )
    cursor = conn.cursor()
except Exception as e:
    print(f"<p>Database connection error: {e}</p>")
    exit()

def get_book_by_keyword(k, cursor):
    """
    Find book data by keyword, handling abbreviations, misspellings, and variations.
    Returns a dictionary with book id, name, chapters, and the numeric reference.
    """
    # Initialize return dictionary
    return_data = {
        'id': 0,
        'book': '',
        'chapters': 0,
        'num_key': '',
        'queryText': ''
    }
    
    # Clean the input
    _k = unquote(k)
    # Remove extra spaces
    _k = re.sub(r'(\s){2,}', ' ', _k)
    # Remove periods
    _k = _k.replace('.', '')
    
    # Make ordinals regular numbers (expecting space between ordinal and book name)
    _k = _k.replace('1st ', '1 ')
    _k = _k.replace('2nd ', '2 ')
    _k = _k.replace('3rd ', '3 ')
    _k = _k.replace('first ', '1 ')
    _k = _k.replace('second ', '2 ')
    _k = _k.replace('third ', '3 ')
    _k = re.sub(r'^i ', '1 ', _k, flags=re.IGNORECASE)
    _k = re.sub(r'^ii ', '2 ', _k, flags=re.IGNORECASE)
    _k = re.sub(r'^iii ', '3 ', _k, flags=re.IGNORECASE)
    
    # Catch messed up beginning numbers (1pe or 2jn format)
    if re.match(r'^[1-3][a-zA-Z]', _k):
        _k = re.sub(r'^1', '1 ', _k)
        _k = re.sub(r'^2', '2 ', _k)
        _k = re.sub(r'^3', '3 ', _k)
    
    # Explode by spaces
    ref_keys = _k.split(' ')
    
    # If first element is empty, remove it
    if ref_keys and not ref_keys[0]:
        ref_keys.pop(0)
    
    # If first element is a number, combine with second to make the book key
    if len(ref_keys) > 1 and re.match(r'^[1-9]', ref_keys[0]):
        book_key = ref_keys[0] + ' ' + ref_keys[1]
        num_keys = ref_keys[2] if len(ref_keys) > 2 else ''
    else:
        book_key = ref_keys[0] if ref_keys else ''
        num_keys = ref_keys[1] if len(ref_keys) > 1 else ''
    
    # Add the chapter and verse(s) to the output
    return_data['num_key'] = num_keys
    
    # Handle special cases
    if book_key.lower() == 'jud':
        book_key = 'Judges'
    if book_key.lower() == 'eph':
        book_key = 'Ephesians'
    
    # Build the query with scoring system
    # Look in abbr column for matches (delimited by |) and score matches
    query_text = """
        SELECT * FROM `kjv_books` 
        WHERE `abbr` LIKE CONCAT('%%|', %s, '|%%') 
           OR `book` LIKE %s 
           OR `kjav_abr` LIKE %s 
           OR `book` SOUNDS LIKE %s
        ORDER BY
            CASE WHEN `abbr` LIKE CONCAT('%%|', %s, '|%%') THEN 4 ELSE 0 END
          + CASE WHEN `book` LIKE %s THEN 3 ELSE 0 END
          + CASE WHEN `kjav_abr` LIKE %s THEN 2 ELSE 0 END
          + CASE WHEN `book` SOUNDS LIKE %s THEN 1 ELSE 0 END
        DESC LIMIT 1
    """
    
    try:
        cursor.execute(query_text, (book_key, book_key, book_key, book_key,
                                     book_key, book_key, book_key, book_key))
        result = cursor.fetchone()
        
        if result:
            # Handle Psalms special case
            if result['book'] == 'Psalms':
                result['book'] = 'Psalm'
            
            return_data['id'] = result['id']
            return_data['book'] = result['book']
            return_data['chapters'] = result['chapters']
            return_data['dbRow'] = result
        
        return_data['queryText'] = cursor.mogrify(query_text, (book_key, book_key, book_key, book_key,
                                                                 book_key, book_key, book_key, book_key))
    except Exception as e:
        print(f"<p>Query error: {e}</p>")
    
    return return_data


# Test references array
references = {
    1: 'Jn 3:16',
    2: '2 pt 1:1',
    3: '2nd pt 1:1',
    4: 'ii pt 1:1',
    5: 'Second Peter 1:1',
    6: 'gnesis 1:1'
}

notes = {
    1: 'A typical abbreviation',
    2: 'Several examples of how books with numbers are entered. Just a number',
    3: '...Number with ordinal',
    4: '...Roman numeral',
    5: '...Ordinal word',
    6: 'Common misspelling'
}

print("<html><head><title>Catch Various Book Reference Types</title></head><body>")


# Process each reference
for n, reference in references.items():
    # Reserve the original reference
    _reference = reference
    
    # Get the note to display
    note = notes[n]
    
    # Make all book references lower case to avoid confusing the filter
    reference_lower = reference.lower()
    
    # Get book data using the function
    book_data = get_book_by_keyword(reference_lower, cursor)
    
    # Get the book id for the next query
    bid = book_data['id']
    book = book_data['book']
    
    # Split the chapter from the verse
    if ':' in book_data['num_key']:
        chapter, verse = book_data['num_key'].split(':')
        
        # Build the query for the verse text
        verse_query = """
            SELECT `text` FROM `kjvs`
            WHERE `book` = %s
            AND `chapter` = %s
            AND `verse` = %s
            LIMIT 1
        """
        
        try:
            cursor.execute(verse_query, (bid, chapter, verse))
            verse_result = cursor.fetchone()
            
            if verse_result:
                text = verse_result['text']
                
                # Filter out Strong's numbers (we don't need them now)
                text = re.sub(r'\{(.*?)\}', '', text)
                
                # Display the result
                print(f"Original reference: {_reference} ({note})<br>")
                print(f"Result: &ldquo;{text}&rdquo;&mdash;{book} {chapter}:{verse}<hr>")
            else:
                print(f"Original reference: {_reference} ({note})<br>")
                print(f"No verse found for {book} {chapter}:{verse}<hr>")
        except Exception as e:
            print(f"<p>Error retrieving verse: {e}</p>")
    else:
        print(f"Original reference: {_reference} ({note})<br>")
        print(f"Could not parse chapter:verse from '{book_data['num_key']}'<hr>")

print("</body></html>")

# Close database connection
cursor.close()
conn.close()